rfd/00xx-snowflake-support.md

```diff
@@ -0,0 +1,171 @@
+
+### Teleport configuration
+
+Like in case of other databases, a user will have to add an entry to the teleport configuration file. Example:
+
+```yaml
+db_service:
+  enabled: "yes"
+  databases:
+    - name: "snowflake"
+      description: "❄ Snowflake database"
+      protocol: "snowflake"
+      uri: "im12345.us-east-2.aws.snowflakecomputing.com"
+```
+
+The next step would be to add teleport CA to the Snowflake user account. `tctl` could learn to export the required keys 
+in the same way as it does today for different databases (`tctl auth sign` command). The exported key must then be added 
+to Snowflake by an administrator. The procedure of adding the key to Snowflake is described [here](https://docs.snowflake.com/en/user-guide/key-pair-auth.html#step-4-assign-the-public-key-to-a-snowflake-user).
+
+### Database access support
+
+`tsh db login` should export the user private key currently stored for example in 
+`~/.tsh/keys/proxy.example.com/alice` in PCKS8 format (that's the only format supported by `snowsql` and some SDKs).
+That key should be used to sign a "Teleport JWT". Client then should send the request with generated JWT to authenticate 
+with Snowflake. Teleport can intercept HTTP communication between a client and the Snowflake. The first request send by 
+the client should be to the `/session/v1/login-request` endpoint which contains all login information as username, JWT. 
+This data can be used to authenticate user in Teleport. JWT token prepared by the client can be then read and replaces by Snowflake JWT
+generated by Teleport DB service. Teleport should use Database CA to sign JWT at that point to sign the request instead 
+of User CA to be able to authenticate with the key generated by `tctl auth sign` and imported to Snowflake. 
+In this way the real token is not accessible by a user and Teleport can still validate if the request 
+comes from a trusted source.
+
+```mermaid
+sequenceDiagram
+    participant client
+    participant Teleport
+    participant Snowflake
+    client->>Teleport: HTTP /session/v1/login-request
+    Teleport->>Teleport: RBAC check, genererate and replace JWT
+    Teleport->>Snowflake: HTTP /session/v1/login-request
+    Snowflake->>Teleport: Authentication Token
+    Teleport->>Teleport: Generate Teleport Auth token
+    Teleport->>client: Send Teleport Auth token
+```
+
+On every request Teleport can replace the "Teleport Auth token" with Snowflake Auth token:
+
+```mermaid
+sequenceDiagram
+    participant client
+    participant Teleport
+    participant Snowflake
+    client->>Teleport: HTTP /queries/v1/query-request
+    Teleport->>Teleport: Replace Auth header
+    Teleport->>Snowflake: HTTP /queries/v1/query-request
+    Snowflake->>Teleport: Response
+    Teleport->>client: Forward the response
+```
+
+With this approach the real session token never leaves the Teleport and any client is not able to bypass Teleport
+by extracting the session token.
```

Comment:
Wonder about situation where we have 2  or more HA Snowflake agent. Does the API allows to generate and uses `n` tokens simultaneously ? If not we should store and sync token generation between HA Snowflake agents.

Reply:
Good point. I noticed the problem when I started writing the implementation. I think there is a few scenarios related to tokens:
1. Single connection per client application, single DB proxy (ex. some GUI clients).
2. Multiple connections per client application, single DB proxy (ex.`snowsql` uses 4 connections by default).
3. Multiple connections per client application, DB proxy in HA configuration.

We need at least one token per connected user. I think that we can assume that we can get one token per connected client as I don't think that sharing tokens between multiple users is a good idea. API is not documented (at least I wasn't able to find any official docs), so I'm not sure what limits apply (how many tokens can we generate, what is the rate limiting, etc).
The simplest implementation would generate one token per connection and keep it in memory. Connection comes from the client, we generate a new token. That works fine for clients that create only one connection, as client using multiple connections would generate multiple tokens. We would generate more token than we need. Each connection may go to different DB agent, so we would need to share the tokens if we don't want to generate extra tokens. That would complicate the implementation tho. Let's assume that you're using `snowsql`. At the beginning `snowsql` creates one connection to authenticate the client. Right after the successful response is returned `snowsql` creates additional connections using the generated token. Sharing the token would have to be quick, because new connection can go to another DB agent that doesn't have the token generated by different DB agent. We could use some kind of "sticky session" to keep all connection from a single machine routed to the same DB proxy or we could encrypt the access token and return it back to the client. That would simplify the flow on our side. That would also help us with token rotation as client needs to request the rotation anyway, so we could rotate the token only then. We would just need to make sure that the encrypted token returned to a client is secure 😅 

TLDR:
1. Create one token per connection. 
Pros: simple implementation
Cons: we can potentially hit some limits as we don't know what they are. We also need to rotate each token in the background.
2. Return encrypted token to the client
Pros: Should work in any configuration and reduces the number of tokens we generate
Cons: Potential security risk if we implement the encryption wrong.